cirq/circuits/frozen_circuit_test.py

```diff
@@ -0,0 +1,2780 @@
+from random import randint, random, sample, randrange
+import os
+import numpy as np
+import pytest
+import sympy
+
+import cirq
+import cirq.google as cg
+import cirq.testing
+from cirq import ops
+
+
+class _MomentAndOpTypeValidatingDeviceType(cirq.Device):
+
+    def validate_operation(self, operation):
+        if not isinstance(operation, cirq.Operation):
+            raise ValueError('not isinstance({!r}, {!r})'.format(
+                operation, cirq.Operation))
+
+    def validate_moment(self, moment):
+        if not isinstance(moment, cirq.Moment):
+            raise ValueError('not isinstance({!r}, {!r})'.format(
+                moment, cirq.Moment))
+
+
+moment_and_op_type_validating_device = _MomentAndOpTypeValidatingDeviceType()
+
+
+def test_equality():
+    a = cirq.NamedQubit('a')
+    b = cirq.NamedQubit('b')
+
+    eq = cirq.testing.EqualsTester()
+
+    # Default is empty. Iterables get listed.
+    eq.add_equality_group(cirq.FrozenCircuit(),
+                          cirq.FrozenCircuit(device=cirq.UNCONSTRAINED_DEVICE),
+                          cirq.FrozenCircuit([]), cirq.FrozenCircuit(()))
+    eq.add_equality_group(cirq.FrozenCircuit([cirq.Moment()]),
+                          cirq.FrozenCircuit((cirq.Moment(),)))
+    eq.add_equality_group(cirq.FrozenCircuit(device=cg.Foxtail))
+
+    # Equality depends on structure and contents.
+    eq.add_equality_group(cirq.FrozenCircuit([cirq.Moment([cirq.X(a)])]))
+    eq.add_equality_group(cirq.FrozenCircuit([cirq.Moment([cirq.X(b)])]))
+    eq.add_equality_group(
+        cirq.FrozenCircuit([cirq.Moment([cirq.X(a)]),
+                            cirq.Moment([cirq.X(b)])]))
+    eq.add_equality_group(
+        cirq.FrozenCircuit([cirq.Moment([cirq.X(a), cirq.X(b)])]))
+
+    # Big case.
+    eq.add_equality_group(
+        cirq.FrozenCircuit([
+            cirq.Moment([cirq.H(a), cirq.H(b)]),
+            cirq.Moment([cirq.CZ(a, b)]),
+            cirq.Moment([cirq.H(b)]),
+        ]))
+    eq.add_equality_group(
+        cirq.FrozenCircuit([
+            cirq.Moment([cirq.H(a)]),
+            cirq.Moment([cirq.CNOT(a, b)]),
+        ]))
+
+
+def test_approx_eq():
+
+    class TestDevice(cirq.Device):
+
+        def validate_operation(self, operation: cirq.Operation) -> None:
+            pass
+
+    a = cirq.NamedQubit('a')
+    b = cirq.NamedQubit('b')
+
+    assert not cirq.approx_eq(cirq.FrozenCircuit([cirq.Moment([cirq.X(a)])]),
+                              cirq.Moment([cirq.X(a)]))
+
+    assert cirq.approx_eq(cirq.FrozenCircuit([cirq.Moment([cirq.X(a)])]),
+                          cirq.FrozenCircuit([cirq.Moment([cirq.X(a)])]))
+    assert not cirq.approx_eq(cirq.FrozenCircuit([cirq.Moment([cirq.X(a)])]),
+                              cirq.FrozenCircuit([cirq.Moment([cirq.X(b)])]))
+
+    assert cirq.approx_eq(
+        cirq.FrozenCircuit([cirq.Moment([cirq.XPowGate(exponent=0)(a)])]),
+        cirq.FrozenCircuit([cirq.Moment([cirq.XPowGate(exponent=1e-9)(a)])]))
```

Comment:
Is the `approx` in `approx_eq` referring to types? I don't think a Circuit should register as equal to a FrozenCircuit otherwise - if users want to compare the two, they can `[un]freeze` as necessary.